solana-programs/audius_eth_registry/src/processor.rs

```diff
@@ -132,6 +266,278 @@ impl Processor {
+
+        let instruction_data_3 = Self::recover_instruction_data(&signature_data_3, &valid_signer_3);
+
+        if instruction_data_1 != secp_instruction_1.data {
+            return Err(AudiusError::SignatureVerificationFailed.into());
+        }
+
+        if instruction_data_2 != secp_instruction_2.data {
+            return Err(AudiusError::SignatureVerificationFailed.into());
+        }
+
+        if instruction_data_3 != secp_instruction_3.data {
+            return Err(AudiusError::SignatureVerificationFailed.into());
+        }
+
+        let signer_group = Box::new(SignerGroup::try_from_slice(
+            &signer_group_info.data.borrow(),
+        )?);
+
+        if !signer_group.is_initialized() {
+            return Err(AudiusError::UninitializedSignerGroup.into());
+        }
+
+        // Each signature data message is expected to be a recent unix timestamp
+        // If messages do not adhere to this format, the operation will fail
+        let timestamp_result = Self::validate_timestamp_messages(
+            &clock,
+            &signature_data_1.message,
+            &signature_data_2.message,
+            &signature_data_3.message,
+        );
+
+        if timestamp_result.is_err() {
+            return Err(AudiusError::SignatureVerificationFailed.into());
+        }
+
+        let mut old_valid_signer = Box::new(ValidSigner::try_from_slice(
+            &old_valid_signer_info.data.borrow(),
+        )?);
+
+        if !old_valid_signer.is_initialized() {
+            return Err(AudiusError::ValidSignerNotInitialized.into());
+        }
+
+        if old_valid_signer.signer_group != *signer_group_info.key {
+            return Err(AudiusError::WrongSignerGroup.into());
+        }
+
+        old_valid_signer.version = Self::VALID_SIGNER_UNINITIALIZED_VERSION;
+        old_valid_signer
+            .serialize(&mut *old_valid_signer_info.data.borrow_mut())
+            .map_err(|e| e.into())
+    }
+
+    /// Process [ValidateMultipleSignaturesAddSigner]()
+    pub fn process_multiple_signatures_add_signer(
+        accounts: &[AccountInfo],
+        signature_data_1: SignatureData,
+        signature_data_2: SignatureData,
+        signature_data_3: SignatureData,
+        eth_address: [u8; SecpSignatureOffsets::ETH_ADDRESS_SIZE],
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        // initialized valid signer account 1
+        let valid_signer_1_info = next_account_info(account_info_iter)?;
+        // initialized valid signer account 2
+        let valid_signer_2_info = next_account_info(account_info_iter)?;
+        // initialized valid signer account 3
+        let valid_signer_3_info = next_account_info(account_info_iter)?;
+        // signer group account
+        let signer_group_info = next_account_info(account_info_iter)?;
+        // incoming valid signer account
+        let new_valid_signer_info = next_account_info(account_info_iter)?;
```

Comment:
One thing missing here (and in the initial prototype) is the check for account ownership. For all the valid signers and signer groups the owner should be this program ID. Otherwise others can fool this program by supplying random accounts from other programs.

Reply:
This comment is valid for all the instructions.